File: problem49.py
class Solution:
    def groupAnagrams(self, strs: List[str]):
        # Base cases:
        if len(strs) == 0:
            return [[""]]
        if len(strs) == 1:
            return [[strs[0]]]

        # go through everything 
        # checking and adding in set or dictionary;
        # we can sort them in and check in dictionary and add index if it exists there

        anagrams = {} # hash table for words
        for i in range(len(strs)):
            sorted_str = sorted(strs[i])
            newStr = "".join(sorted_str)
            if newStr in anagrams:
                anagrams[newStr].append(strs[i])
            else:
                anagrams[newStr] = [strs[i]]
        ans = []
        for value in anagrams.values():
            ans.append(value)
        return ans

File: problem49.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <algorithm>
#include <iostream>

using namespace std;

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        // Base case
        if(strs.size() == 0){
            return {{}};
        }

        // hashmap for all words
        unordered_map<string, vector<string>> anagrams;

        for(string s : strs){
            string temp = s; // save word before sort
            sort(s.begin(), s.end()); // sort the word
            anagrams[s].push_back(temp); // push word to it
        }
        vector<vector<string>> result;

        for (auto& pair : anagrams) {
            result.push_back(pair.second);
        }
        return result;
    }
};